### PR TITLE
feat: apply upstream(Optimism) bugfixes

### PR DESCRIPTION
**Pull request by seolaoh**
# Description

Related [[KROMA-004] Upstream (Optimism) bugfixes should be applied](https://github.com/chainlight-io/2023-kroma-network-audit-issues-client/issues/4) reported by theori audit.
It is a bunch of upstream bugfixes of Optimism, and I've added related links in each commit message. I also made related changes to the modified files beyond the reported issue.

**NOTE:** It reflects just a few of the changes since Optimism v1.0.3, the last version we applied.

### CODE REVIEW

**Comment by seolaoh** [comment]
@kroma-network/devops
I figured out the reason why the solidity test is only failed in CI but success in local. The reason is the difference of foundry version. We have replaced the version of foundry in CI in [this commit](https://github.com/kroma-network/kroma/commit/b215664f09166a5a638c0590b68080812f30a285), which is [this version](https://github.com/foundry-rs/foundry/releases/tag/nightly-e15e33a07c0920189fc336391f538c3dad53da73). I think it would be better to upgrade foundry to [this commit version](https://github.com/foundry-rs/foundry/commit/3b1129b5bc43ba22a9bcf4e4323c5a9df0023140). Recently Optimism also upgraded to this version in [this PR](https://github.com/ethereum-optimism/optimism/pull/6153) and changed the way to manage foundry version in [this PR](https://github.com/ethereum-optimism/optimism/pull/6162). Maybe we need to apply it!

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -57,7 +57,7 @@ jobs:
       - name: Install Foundry
         uses: foundry-rs/foundry-toolchain@v1
         with:
-          version: nightly-e15e33a07c0920189fc336391f538c3dad53da73
+          version: nightly-d369d2486f85576eec4ca41d277391dfdae21ba7
```
Comment (seolaoh) [inline .github/workflows/ci.yml:60]: I updated foundry version to [this nightly version](https://github.com/foundry-rs/foundry/releases/tag/nightly-d369d2486f85576eec4ca41d277391dfdae21ba7), which includes [this commit version](https://github.com/foundry-rs/foundry/commit/3b1129b5bc43ba22a9bcf4e4323c5a9df0023140) that Optimism is using.

**Review by Pangssu — APPROVED** [review]
LGTM

File: FORK.md [hunk 2]
```diff
@@ -1,7 +1,26 @@
 # Fork
 
-## optimism
+## Optimism
 
 - tag: `v1.0.3`
 - differences
   - [Differences from Optimism Bedrock](./specs/differences-from-optimism-bedrock.md)
+
+### Others
```
Comment (0xHansLee) [inline FORK.md:9]: I think this `FORK.md` should only include the difference from Optimism.
Why don't we remove these list in `FORK.md`, and leave in this PR's description?

**Review by 0xHansLee — APPROVED** [review]

File: FORK.md (hunk 2 above)
Comment (seolaoh) [reply FORK.md:9]: I added this list to match with [this change](https://github.com/kroma-network/kroma/pull/136/files#diff-93ec59db0ab516baf2c0eebb87cf68bbac382c2d5bff18c1aa8f9abca3ee729d). I agree this list should be removed after applying upstream changes by tag like `v1.0.7`. Wdyt?

Comment (0xHansLee) [reply FORK.md:9]: Ok. Let's remove those list after applying upstream changes.

**Review by kangsorang — APPROVED** [review]
LGTM

**Pull request merged**